src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -90,34 +91,102 @@ checkNoRecursion :: (MonadError String me) => DatalogProgram -> me ()
+addRhsToRules:: RuleRHS -> [Rule] -> [Rule]
+addRhsToRules toAdd rules =
+     map (\r -> Rule (rulePos r) (ruleLHS r) (toAdd : ruleRHS r)) rules
+
+convertFtlStatement :: (MonadError String me) => Statement -> me [Rule]
+convertFtlStatement (ForStatement p e r Nothing s) = do
+    rules <- convertFtlStatement s
+    let atom = Atom p r [("", e)]
+    let rhs = RHSLiteral True atom
+    let rules' = addRhsToRules rhs rules
+    return rules'
+convertFtlStatement (ForStatement p e r (Just c) s) = do
+    rules <- convertFtlStatement s
+    let atom = Atom p r [("", e)]
+    let rhs0 = RHSLiteral True atom
+    let rhs1 = RHSCondition c
+    let rules' = map (\r -> Rule (rulePos r) (ruleLHS r) (rhs0 : rhs1 : ruleRHS r)) rules
+    return rules'
+convertFtlStatement (IfStatement p c s Nothing) = do
+    rules <- convertFtlStatement s
+    let rules' = addRhsToRules (RHSCondition c) rules
+    return rules'
+convertFtlStatement (IfStatement p c s (Just e)) = do
+    rules0 <- convertFtlStatement s
+    rules1 <- convertFtlStatement e
+    let rules0' = addRhsToRules (RHSCondition c) rules0
+    let rules1' = addRhsToRules (RHSCondition $ eNot c) rules1
+    return (rules0' ++ rules1')
+convertFtlStatement (MatchStatement p e c) = do
+    rulesList <- mapM (\(co, s) -> convertFtlStatement s) c
+    let matchList = explodeMatchCases $ map fst c
+    let matchExpressions = map (\l -> RHSCondition $ eMatch e l) matchList
+    let rules = concat $ zipWith addRhsToRules matchExpressions rulesList
+    return rules
+convertFtlStatement (LetStatement p l s) = do
+    rules <- convertFtlStatement s
+    let exprs = map (\a -> eSet (eVarDecl $ leftAssign a) (rightAssign a)) l
+    let rhs = map (\e -> RHSCondition e) exprs
+    let rules' = map (\r -> Rule (rulePos r) (ruleLHS r) ((ruleRHS r) ++ rhs)) rules
+    return rules'
+convertFtlStatement (BlockStatement p l) = do
+    rulesList <- mapM convertFtlStatement l
+    let rules = concat rulesList
+    return rules
+convertFtlStatement (EmptyStatement p) =
+    do return [Rule p [] []]
+convertFtlStatement (InsertStatement p r v) = do
+    let atom = Atom p r $ map (\e -> ("", e)) v
+    return [Rule p [atom] []]
+
 -- Remove syntactic sugar
 progDesugar :: (MonadError String me) => DatalogProgram -> me DatalogProgram
-progDesugar d0 = do 
+progDesugar d0 = do
     d1 <- progDesugarAtoms d0
     progExprMapCtxM d1 (exprDesugar d1)
- 
--- Desugar atoms: convert all relational atoms to named field syntax.  
+
+-- Desugar atoms: convert all relational atoms to named field syntax.
 progDesugarAtoms :: (MonadError String me) => DatalogProgram -> me DatalogProgram
 progDesugarAtoms d = do
-    rules' <- mapM (\r -> do lhs <- mapM (atomDesugarArgs d) $ ruleLHS r
-                             rhs <- mapM (rhsDesugarAtoms d) $ ruleRHS r 
-                             return r{ruleLHS = lhs, ruleRHS = rhs})
-                   $ progRules d
-    return d{progRules = rules'}    
+    -- Drop rules with an empty LHS
```

Comment:
It would be cleaner if the function that introduces rules with empty LHS filtered them out before adding them to the spec.